lib/src/gzip_encoding.dart

```diff
@@ -0,0 +1,230 @@
+import 'dart:async';
+import 'dart:io' show HttpHeaders, ZLibEncoder;
+import 'dart:math' as math;
+import 'dart:typed_data';
+
+import 'package:shelf/shelf.dart';
+
+/// Using level 4, since have almost the same compression ratio of normal
+/// texting content, but with a lower CPU usage. This is the recommended
+/// for live data compression with `gzip`.
+const int defaultGzipEncodingCompressionLevel = 4;
+
+final _defaultGzipEncoder =
+    ZLibEncoder(gzip: true, level: defaultGzipEncodingCompressionLevel);
+
+/// The default `gzip` encoding [Middleware].
+final Middleware gzipMiddleware = createGzipMiddleware();
+
+/// Converts a [Response] to a `gzip` encoding
+/// (only if the [Request] [acceptsGzipEncoding]).
+///
+/// - The [compressionLevel] for the `gzip` encoder. Default: 4
+/// - If [addCompressionRatioHeader] is `true`, add header `X-Compression-Ratio`.
+Middleware createGzipMiddleware(
+    {int compressionLevel = defaultGzipEncodingCompressionLevel,
+    bool addCompressionRatioHeader = true}) {
+  return (Handler innerHandler) {
+    return (request) {
+      if (!acceptsGzipEncoding(request)) {
+        return innerHandler(request);
+      }
+      return Future.sync(() => innerHandler(request)).then((response) =>
+          gzipEncodeResponse(response,
+              compressionLevel: compressionLevel,
+              addCompressionRatioHeader: addCompressionRatioHeader));
+    };
+  };
+}
+
+/// Returns `true` if the request accepts `gzip` encoding.
+bool acceptsGzipEncoding(Request request) {
+  var acceptEncoding = request.headers[HttpHeaders.acceptEncodingHeader];
+  return acceptEncoding?.contains('gzip') ?? false;
+}
+
+/// Converts [response] to a `gzip` encoding response.
+/// Checks [canGzipEncodeResponse].
+///
+/// See [createGzipMiddleware].
+FutureOr<Response> gzipEncodeResponse(Response response,
```

Comment:
make private?

Reply:
I think that the developer can use this if he wants to create his own `Middleware` and also provide Gzip in the same `Middleware`, without need to reimplement everything that is in this package. We have a internal case that needs that.

Reply:
okay, that makes sense.